micrometer-core/src/main/java/io/micrometer/core/instrument/distribution/TimeWindowFixedBoundaryHistogram.java

```diff
@@ -116,6 +117,31 @@ void resetAccumulatedHistogram() {
                 : currentHistogram().countAtValue(value);
     }
 
+    @Override
+    Iterator<CountAtBucket> countAtValues(Iterator<Double> values) {
+        return new Iterator<CountAtBucket>() {
+            private double accumulation = 0.0;
+
+            @Override
+            public boolean hasNext() {
+                return values.hasNext();
+            }
+
+            @Override
+            public CountAtBucket next() {
+                double value = values.next();
+                double count = currentHistogram().countAtValue(value);
+                if (cumulativeBucketCounts) {
+                    accumulation += count;
+                    return new CountAtBucket(value, accumulation);
```

Comment:
Is overriding this method here needed? Does the `countAtValue` implementation above not take care of the logic here?

Reply:
@shakuzen I think the problem is that the cumulative count is recomputed for each bucket.

For example, imagine if there are bucket boundaries for values 4.0 and 5.0 and the cumulative count less or equal to 4.0 has just been computed as part of `takeCountSnapshot`:
https://github.com/micrometer-metrics/micrometer/blob/994ec8bb8bea96ab072829aa50754c6cf4a3e69f/micrometer-core/src/main/java/io/micrometer/core/instrument/distribution/AbstractTimeWindowHistogram.java#L215 Then, a new value of 1.0 comes in before computing the cumulative count less or equal to 5.0. The count <= 5.0 will be greater than the count <= 4.0 even though no such value has been seen.

Does that make sense? You're probably right that a fix would need to be added for `+Inf` as well, perhaps including it in the histogram?

Reply:
Thank you for walking me through it. That makes sense. What I missed in my initial reading of the changes is that line 133 is calling `currentHistogram().countAtValue(value)` rather than `countAtValue(value)`. This effectively makes the method `countAtValueCumulative` never called anymore, and it feels like the code should be further reworked to reflect that. The proposed approach is more efficient algorithmically, and it avoids an inconsistency. It's a step in the right direction. The snapshot still isn't atomic, though. Going off your example, if there were also a 10.0 bucket and the value recorded during taking the snapshot were 7, then the 10.0 bucket is going to have a count 1 higher than when the snapshot started being taken.

I wrote a jcstress test to capture this in https://github.com/micrometer-metrics/micrometer/pull/3396. With the changes proposed here, failure cases are reduced but not entirely eliminated.

> for `+Inf` as well, perhaps including it in the histogram?

If the buckets do not already include infinity, it will be added using the `count`. The current logic for the infinity bucket is: https://github.com/micrometer-metrics/micrometer/blob/994ec8bb8bea96ab072829aa50754c6cf4a3e69f/implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheus/PrometheusMeterRegistry.java#L267-L279
There's some discussion on it https://github.com/micrometer-metrics/micrometer/issues/998 previously.

Reply:
> This effectively makes the method countAtValueCumulative never called anymore, and it feels like the code should be further reworked to reflect that.

I removed the now unused methods.

> It's a step in the right direction. The snapshot still isn't atomic, though.

I believe our examples have slightly different properties though. This pull request only fixes the computation of cumulative counts so that they don't exhibit counts for values that were never seen. It indeed does not make the computation atomic.

Making that computation atomic would require synchronization primitives which would affect performance. The current consistency - performance trade-off could be sufficient for metrics, it is correct for some reordering of operations.